Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp

```diff
@@ -515,6 +515,60 @@ bool RemoteRenderingBackendProxy::isCached(const ImageBuffer& imageBuffer) const
     return false;
 }
 
+#if USE(GRAPHICS_LAYER_WC)
+
+class RemoteRenderingBackendProxyFlushFence : public ThreadSafeRefCounted<RemoteRenderingBackendProxyFlushFence> {
+    WTF_MAKE_NONCOPYABLE(RemoteRenderingBackendProxyFlushFence);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static Ref<RemoteRenderingBackendProxyFlushFence> create(IPC::Semaphore semaphore)
+    {
+        return adoptRef(*new RemoteRenderingBackendProxyFlushFence { WTFMove(semaphore) });
+    }
+
+    bool waitFor(Seconds timeout)
+    {
+        Locker locker { m_lock };
+        if (m_signaled)
+            return true;
+        m_signaled = m_semaphore.waitFor(timeout);
+        return m_signaled;
+    }
+
+    std::optional<IPC::Semaphore> tryTakeSemaphore()
+    {
+        if (!m_signaled)
+            return std::nullopt;
+        Locker locker { m_lock };
+        return WTFMove(m_semaphore);
+    }
+
+private:
+    RemoteRenderingBackendProxyFlushFence(IPC::Semaphore semaphore)
+        : m_semaphore(WTFMove(semaphore))
+    {
+    }
+    Lock m_lock;
+    std::atomic<bool> m_signaled { false };
+    IPC::Semaphore WTF_GUARDED_BY_LOCK(m_lock) m_semaphore;
+};
+
+Function<void()> RemoteRenderingBackendProxy::createFlusher()
+{
+    std::optional<IPC::Semaphore> flushSemaphore;
+    if (m_pendingFlush)
+        flushSemaphore = m_pendingFlush->tryTakeSemaphore();
```

Comment:
A lot of the complexity around recycling the semaphore here was added because creating and serialising the Semaphore (on Cocoa!) was pretty expensive when done per-image. For pages with a large number of (tiled) layers, this ended up being mildly expensive.

This code recycles the Semaphore, but doesn't prevent repeated serialisation.

Given that this is a single object per-frame, I would strongly suspect that the performance doesn't matter, and you could remove `RemoteRenderingBackendProxyFlushFence` and just return a lambda wrapping a newly allocated `Semaphore` each time.

Reply:
Fixed.